### PR TITLE
Fix issue where integer underflow caused exponential backoff time to retry to underflow

### PR DESCRIPTION
**Pull request by timtay-microsoft**
Issue #969

During the exponential backoff algorithm, some double values were cast to integer. When these values were large, the cast would result in an underflow which would cause the recommended next delay to be negative unexpectedly.

~~This pr alters the algorithm to only use doubles (which go to infinity, not to underflow) to fix this issue~~

Switched back to integers, and using checked statements to lookout for integer overflow

### CODE REVIEW

File: iothub/device/tests/ExponentialBackoffTests.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿using Microsoft.Azure.Devices.Client.TransientFaultHandling;
```
Comment (CIPop) [inline iothub/device/tests/ExponentialBackoffTests.cs]: Missing required header.

File: iothub/device/src/TransientFaultHandling/ExponentialBackoff.cs [hunk 2]
```diff
@@ -98,9 +98,9 @@ public override ShouldRetry GetShouldRetry()
                 if (currentRetryCount < this.retryCount)
                 {
                     Random random = new Random();
-                    int num = (int)((Math.Pow(2.0, (double)currentRetryCount) - 1.0) * (double)random.Next((int)(this.deltaBackoff.TotalMilliseconds * 0.8), (int)(this.deltaBackoff.TotalMilliseconds * 1.2)));
-                    int num2 = (int)Math.Min(this.minBackoff.TotalMilliseconds + (double)num, this.maxBackoff.TotalMilliseconds);
-                    retryInterval = TimeSpan.FromMilliseconds((double)num2);
+                    double num = ((Math.Pow(2.0, currentRetryCount) - 1.0) * random.Next((int)(this.deltaBackoff.TotalMilliseconds * 0.8), (int)(this.deltaBackoff.TotalMilliseconds * 1.2)));
```
Comment (CIPop) [inline iothub/device/src/TransientFaultHandling/ExponentialBackoff.cs]: 1. Why do you need to use `double` which is significantly slower and harder to debug. The end result is a TimeSpan which is based on `milliseconds`. I doubt anyone would care of timings less than a millisecond. Instead you can use something like ulong which should be large enough for these purposes. You may also want to use a `checked` statement to ensure no overflows.

2. What does it mean here to have a timespan formed from Double.MinValue, Double.MaxValue or Infinity? That basically makes the device either not wait (spin-wait causing high CPU) or wait forever (hang).

File: iothub/device/tests/ExponentialBackoffTests.cs (hunk 1 above)
Comment (timtay-microsoft) [reply iothub/device/tests/ExponentialBackoffTests.cs]: Sure, I'll add it

File: iothub/device/src/TransientFaultHandling/ExponentialBackoff.cs (hunk 2 above)
Comment (timtay-microsoft) [reply iothub/device/src/TransientFaultHandling/ExponentialBackoff.cs]: 1) I'll make use of a checked block here and keep the numbers as integers then. If overflow occurs, the retry policy will just default to the max backoff.

2) This code does not create TimeSpans with Double.MinValue or Double.MaxValue. It creates TimeSpans within the min backoff and the max backoff

File: iothub/device/src/TransientFaultHandling/ExponentialBackoff.cs [hunk 3]
```diff
@@ -98,10 +98,23 @@ public override ShouldRetry GetShouldRetry()
                 if (currentRetryCount < this.retryCount)
                 {
                     Random random = new Random();
-                    int num = (int)((Math.Pow(2.0, (double)currentRetryCount) - 1.0) * (double)random.Next((int)(this.deltaBackoff.TotalMilliseconds * 0.8), (int)(this.deltaBackoff.TotalMilliseconds * 1.2)));
-                    int num2 = (int)Math.Min(this.minBackoff.TotalMilliseconds + (double)num, this.maxBackoff.TotalMilliseconds);
-                    retryInterval = TimeSpan.FromMilliseconds((double)num2);
-                    return true;
+                    try
```
Comment (davilu) [inline iothub/device/src/TransientFaultHandling/ExponentialBackoff.cs]: The code is too complicated to read. How about `if(num2<0) return maxBackoff;`

Comment (timtay-microsoft) [reply iothub/device/src/TransientFaultHandling/ExponentialBackoff.cs]: Overflow may loop back around to look like a normal positive value again, so we do need the checked statement to correctly implement the exponential backoff.

Comment (davilu) [reply iothub/device/src/TransientFaultHandling/ExponentialBackoff.cs]: Then don't convert it to int. 
```
double interval = Math.Pow(2.0, currentRetryCount) - 1.0) * random.Next((int)(this.deltaBackoff.TotalMilliseconds * 0.8), (int)(this.deltaBackoff.TotalMilliseconds * 1.2) + this.minBackoff.TotalMilliseconds;
if (interval > this.maxBackoff.TotalMilliseconds) {
   retryInterval = maxBackoff;
} else [
  retryInterval = TimeSpan.FromMilliseconds(interval);
}
```

Comment (timtay-microsoft) [reply iothub/device/src/TransientFaultHandling/ExponentialBackoff.cs]: As Cristian mentioned above, double's are unnecessarily heavy here. We can keep everything as an int as long as we have the checked statement

**Pull request merged**